File: src/dependencies/crazyflie_hardware/src/crazyflie_hardware_cpp/src/crtp_driver_cpp/localization.cpp
#include "crazyflie_hardware_cpp/crtp_driver_cpp/localization.hpp"
using std::placeholders::_1;
using namespace std::chrono_literals;

Localization::Localization(std::shared_ptr<rclcpp_lifecycle::LifecycleNode> node, CrtpLink *link, std::string tf_name)
    : LocalizationLogic(link)
    , node(node)
    , logger_name(node->get_name())
    , tf_name(tf_name)
    , is_beeing_tracked(false)
    , is_beeing_broadcasted(false)
    , channel_(80)
    , data_rate_(2)
{
    callback_group = node->create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);

    RCLCPP_DEBUG(rclcpp::get_logger(logger_name), "Localization initialized");
}

bool Localization::stop_external_tracking() 
{
    bool ret = true; 
    if (is_beeing_tracked) {
        if (!remove_from_tracker()) ret = false; 
    }
    if (is_beeing_broadcasted)
    {
        if (!remove_from_broadcaster()) ret = false;
    } 
    return true;
}

bool Localization::start_external_tracking(int marker_configuration_index,
                                           int dynamics_configuration_index,
                                           double max_initial_deviation,
                                           std::vector<double> initial_position,
                                           int channel,
                                           int datarate)
{
    is_beeing_tracked = add_to_tracker(marker_configuration_index, dynamics_configuration_index, max_initial_deviation, initial_position);
    if (is_beeing_tracked)
    {
        is_beeing_broadcasted = add_to_broadcaster(channel, datarate);
        if (is_beeing_broadcasted) 
        {
            return true;
        }
    }
    return false;
}

bool Localization::add_to_tracker(
    int marker_configuration_index,
    int dynamics_configuration_index,
    double max_initial_deviation,
    std::vector<double> initial_position)
{
    rclcpp::Client<object_tracker_interfaces::srv::AddTrackerObject>::SharedPtr client;
    if (auto shared_node = node.lock()) {
        client = shared_node->create_client<object_tracker_interfaces::srv::AddTrackerObject>(
            "/tracker/add_object",
            rclcpp::QoS(rclcpp::KeepLast(1)).get_rmw_qos_profile(),
            callback_group);

    } else return false;
    
    if (!client->wait_for_service(1s))
    {
        RCLCPP_WARN(rclcpp::get_logger(logger_name), "Tracking Service not available!");
        return false;
    }

    auto request = std::make_shared<object_tracker_interfaces::srv::AddTrackerObject::Request>();
    request->tf_name.data = tf_name;
    request->marker_configuration_idx = marker_configuration_index;
    request->dynamics_configuration_idx = dynamics_configuration_index;
    request->max_initial_deviation = max_initial_deviation;
    request->initial_pose.position.x = initial_position[0];
    request->initial_pose.position.y = initial_position[1];
    request->initial_pose.position.z = initial_position[2];
    auto result = client->async_send_request(request);

    auto status = result.wait_for(3s); // not spinning here!
    if (status == std::future_status::ready)
    {
        RCLCPP_DEBUG(rclcpp::get_logger(logger_name), "Service call success!");
        auto res = result.get();
        if (res->success)
        {
            RCLCPP_DEBUG(rclcpp::get_logger(logger_name), "Add to tracker success!");
            return true;
        }
        else
        {
            RCLCPP_INFO(rclcpp::get_logger(logger_name), "Add to tracker failed!");
        }
    }
    else
    {
        RCLCPP_DEBUG(rclcpp::get_logger(logger_name), "Service call timed out!");
    }

    return false;
}



bool Localization::add_to_broadcaster(int channel, int datarate)
{
    channel_ = channel;
    data_rate_ = datarate;
    rclcpp::Client<broadcaster_interfaces::srv::PosiPoseBroadcastObject>::SharedPtr client;
    if (auto shared_node = node.lock()) {
        client = shared_node->create_client<broadcaster_interfaces::srv::PosiPoseBroadcastObject>(
                "/add_posi_pose_object",
                rclcpp::QoS(rclcpp::KeepLast(1)).get_rmw_qos_profile(),
                callback_group);
    } else return false;

    if (!client->wait_for_service(1s))
    {
        RCLCPP_WARN(rclcpp::get_logger(logger_name), "Broadcast Service not available!");
        return false;
    }

    auto request = std::make_shared<broadcaster_interfaces::srv::PosiPoseBroadcastObject::Request>();
    request->channel = channel;
    request->data_rate = datarate;
    request->tf_frame_id = tf_name;

    auto result = client->async_send_request(request);

    auto status = result.wait_for(3s); // not spinning here!
    if (status == std::future_status::ready)
    {
        auto res = result.get();
        if (res->success)
        {
            RCLCPP_DEBUG(rclcpp::get_logger(logger_name), "Broadcast success!");
            return true;
        }
        else
        {
            RCLCPP_INFO(rclcpp::get_logger(logger_name), "Add to Broadcast failed!");
        }
    }
    else
    {
        RCLCPP_DEBUG(rclcpp::get_logger(logger_name), "Service call timed out!");
    }

    return false;
}

bool Localization::remove_from_tracker() 
{
    rclcpp::Client<object_tracker_interfaces::srv::RemoveTrackerObject>::SharedPtr client;
    if (auto shared_node = node.lock()) {
         client = shared_node->create_client<object_tracker_interfaces::srv::RemoveTrackerObject>(
                "/tracker/remove_object",
                rclcpp::QoS(rclcpp::KeepLast(1)).get_rmw_qos_profile(),
                callback_group);
    } else return false;

    if (!client->wait_for_service(100ms))
    {
        RCLCPP_WARN(rclcpp::get_logger(logger_name), "Tracking Service not available!");
        return false;
    }
    
    auto request = std::make_shared<object_tracker_interfaces::srv::RemoveTrackerObject::Request>();
    request->tf_name.data = tf_name;
    
    auto result = client->async_send_request(request);
    auto status = result.wait_for(100ms);

    bool ret = status == std::future_status::ready;
    if (!ret) {
        RCLCPP_WARN(rclcpp::get_logger(logger_name), "Tracker didnt respond in time!");
    }
    return ret;
}

bool Localization::remove_from_broadcaster()
{
    rclcpp::Client<broadcaster_interfaces::srv::PosiPoseBroadcastObject>::SharedPtr client;
    if (auto shared_node = node.lock()) {
        client = shared_node->create_client<broadcaster_interfaces::srv::PosiPoseBroadcastObject>(
            "/remove_posi_pose_object",
            rclcpp::QoS(rclcpp::KeepLast(1)).get_rmw_qos_profile(),
            callback_group);
    } else return false;

    if (!client->wait_for_service(100ms))
    {
        RCLCPP_WARN(rclcpp::get_logger(logger_name), "Broadcast Service not available!");
        return false;
    }
    
    auto request = std::make_shared<broadcaster_interfaces::srv::PosiPoseBroadcastObject::Request>();
    request->channel = channel_;
    request->data_rate = data_rate_;
    request->tf_frame_id = tf_name;
    
    auto result = client->async_send_request(request);
    auto status = result.wait_for(100ms);
    bool ret = status == std::future_status::ready;
    if (!ret) {
        RCLCPP_WARN(rclcpp::get_logger(logger_name), "Broadcaster didnt respond in time!");
    }
    return ret;
}

File: src/dependencies/crazyflie_hardware/src/crazyflie_hardware/crtp_driver/localization.py
import rclpy
from rclpy.node import Node
from rclpy import Future
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup

import rclpy.time
from tf2_ros import TransformException
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener

from .crtp_link_ros import CrtpLinkRos
from .crtp_packer_ros import CrtpPackerRos
from crtp.logic.localization_logic import LocalizationLogic

from broadcaster_interfaces.srv import PosiPoseBroadcastObject

from typing import List


class LocalizationException(Exception):
    pass


class Localization(LocalizationLogic):
    def __init__(self, node: Node, crtp_link: CrtpLinkRos, tf_name: str):
        super().__init__(CrtpPackerRos, crtp_link)
        self.node: Node = node
        self.callback_group = MutuallyExclusiveCallbackGroup()
        self.tf_name = tf_name

    #   self.state = 0

    # def start_selflocalization(self, tf_name):
    #    """It is also possible to not use broadcaster for sending external pose.
    #    This does this but is not fully implemented
    #    Args:
    #        tf_name (_type_): _description_
    #    """
    #
    #    self.tf_buffer = Buffer()
    #    self.tf_listener = TransformListener(self.tf_buffer, self.node)
    #    self.timer = self.node.create_timer(
    #        0.2, lambda: self.on_timer(tf_name), callback_group=self.callback_group
    #    )
    #
    # def on_timer(self, tf_name: str):
    #    if self.state == 1:
    #        return
    #    try:
    #        t = self.tf_buffer.lookup_transform("world", tf_name, rclpy.time.Time())
    #    except TransformException as ex:
    #        self.node.get_logger().info("Tracker no Frame")
    #        return
    #
    #    pos = [
    #        t.transform.translation.x * 1,
    #        t.transform.translation.y * 1,
    #        t.transform.translation.z * 1,
    #    ]
    #    self.send_extpos(pos)

    def start_external_tracking(
        self,
        marker_configuration_index: int,
        dynamics_configuration_index: int,
        max_initial_deviation: float,
        initial_position: List[float],
        channel: int,
        datarate: int,
    ) -> bool:
        self.channel: int = channel
        self.datarate: int = datarate

        tracker_success = self.add_to_tracker(
            marker_configuration_index,
            dynamics_configuration_index,
            max_initial_deviation,
            initial_position,
        )
        if not tracker_success:
            raise LocalizationException("Adding to tracker failed.")

        broadcaster_success = self.add_to_broadcaster(channel, datarate)
        if not broadcaster_success:
            self.remove_from_tracker()
            raise LocalizationException("Adding to Broadcaster failed.")

        return True

    def stop_external_tracking(self) -> List:
        tracker_fut = self.remove_from_tracker()
        broadcaster_fut = self.remove_from_broadcaster(self.channel, self.datarate)
        return [tracker_fut, broadcaster_fut]

    def add_to_tracker(
        self,
        marker_configuration_index: int,
        dynamics_configuration_index: int,
        max_initial_deviation: float,
        initial_position: List[float],
    ):
        # Establish Tracking
        from object_tracker_interfaces.srv import AddTrackerObject, RemoveTrackerObject

        add_to_tracker_service = self.node.create_client(
            srv_type=AddTrackerObject,
            srv_name="/tracker/add_object",
            callback_group=self.callback_group,
        )

        if not add_to_tracker_service.wait_for_service(timeout_sec=1.0):
            raise TimeoutError("Add to Tracker Service not available!")

        req = AddTrackerObject.Request()
        req.tf_name.data = self.tf_name
        req.marker_configuration_idx = marker_configuration_index
        req.dynamics_configuration_idx = dynamics_configuration_index
        req.max_initial_deviation = max_initial_deviation
        (
            req.initial_pose.position.x,
            req.initial_pose.position.y,
            req.initial_pose.position.z,
        ) = initial_position
        fut: Future = add_to_tracker_service.call_async(req)
        rclpy.spin_until_future_complete(
            node=self.node, future=fut, executor=self.node.executor, timeout_sec=1.0
        )
        try:
            response: AddTrackerObject.Response = fut.result()
        except:
            raise TimeoutError("Add to Tracker Service did not respond in time.")
        return response.success

    def remove_from_tracker(self):
        from object_tracker_interfaces.srv import RemoveTrackerObject

        remove_from_tracker_service = self.node.create_client(
            srv_type=RemoveTrackerObject,
            srv_name="/tracker/remove_object",
            callback_group=self.callback_group,
        )

        req = RemoveTrackerObject.Request()
        req.tf_name.data = self.tf_name
        return remove_from_tracker_service.call_async(req)

    def add_to_broadcaster(self, channel: int, datarate: int):
        # Establish Broadcasting
        self.add_to_broadcaster_service = self.node.create_client(
            srv_type=PosiPoseBroadcastObject,
            srv_name="/add_posi_pose_object",
            callback_group=self.callback_group,
        )
        if not self.add_to_broadcaster_service.wait_for_service(timeout_sec=1.0):
            raise TimeoutError("Add to Broadcaster Service not available!")

        req = PosiPoseBroadcastObject.Request()
        req.channel = channel
        req.data_rate = datarate
        req.tf_frame_id = self.tf_name
        fut: Future = self.add_to_broadcaster_service.call_async(req)
        rclpy.spin_until_future_complete(
            node=self.node, future=fut, executor=self.node.executor, timeout_sec=1.0
        )
        try:
            response: PosiPoseBroadcastObject.Response = fut.result()
        except:
            raise TimeoutError("Add to Broadcaster Service did not respond.")
        return response.success

    def remove_from_broadcaster(self, channel: int, datarate: int):
        self.remove_from_broadcaster_service = self.node.create_client(
            srv_type=PosiPoseBroadcastObject,
            srv_name="/remove_posi_pose_object",
            callback_group=self.callback_group,
        )
        if self.remove_from_broadcaster_service.wait_for_service(timeout_sec=0.5):
            req = PosiPoseBroadcastObject.Request()
            req.channel = channel
            req.data_rate = datarate
            req.tf_frame_id = self.tf_name
            return self.remove_from_broadcaster_service.call_async(req)
        return None
